### ISSUE TITLE
Release for every feature/bug fix

### ISSUE THREAD

**Issue by peterbe** [mozilla/mozilla-django-oidc#143]
I propose we make releases more frequently. One release for every new feature or bug fix. That way, we don't have have a meeting and say "Ok, what's changed since the last release?" and type that up in `HISTORY.txt` along with a change to `mozilla_django_oidc/__init__.py`'s `__version__` string. 

It is possible to encryt your PyPI password and put that in `.travis.yml` and let travis try to make a PyPI release on every successful build on origin/master. If the version number hasn't change, PyPI will simply respond with a gentle error like "Sorry, that version already exists". 
That means if all you do is commit a change to the documentation or a style fix or a change only to the unit tests, no new PyPI release is made. [For example](https://github.com/peterbe/hashin/blob/master/.travis.yml#L21-L22).

This would require a change of how you make PRs. If your PR indeed contains a feature change or a bug fix, you're *also* expected to edit `HISTORY.txt` and `__init__.py` along with your change. 
If two people make a PR based on version 1.2.3 and they both make PRs that change `HISTORY.txt` and `__init__.py` then it's likely to cause a conflict and the author has to rebase up. 

Another option is to still set up travis to push to PyPI on every origin/master build, but NOT require PRs to contain a change to `HISTORY.txt` and `__init__.py`. Then once your feature/bug fix lands on master you make *another* PR that just edits `HISTORY.txt` and `__init__.py`. 

Thoughts? Ideas? Objections? Endorsements?

**Comment by peterbe** [comment]
Another option is to use git tagging to drive releases. Travis can be set up to only attempt to make a PyPI release when a new tag is pushed to origin/master. [For example](https://github.com/jazzband/django-configurations/blob/master/.travis.yml#L40-L41) 
But you still have to make a code change, [for example](https://github.com/jazzband/django-configurations/commit/376b5947a8455e3f1bd01dd2dbc17c79cc7aaef1), and *then* make a tag as well. So it's more work which is why I'm less convinced about it personally. 

Perhaps someone can remind us the advantages of "making releases in the source code" **and** making tags.

**Comment by johngian** [comment]
I agree with the idea of having faster releases. Although I agree that the first approach looks better I generally like projects with github releases in sync with pypi releases.

Maybe we can combine both:

* Minor fixes without changes in ``HISTORY.txt`` and ``__init__.__version__`` don't trigger a build
* Features should always be accompanied with changelog and version bump
* Travis will build and push PyPI releases from master
* On successful PyPI release travis will publish a new github release.

Thoughts?

**Comment by johngian** [comment]
I opened a PR with the initial proposal of @peterbe. Apparently travis isn't really flexible to chain a github deployment on succesful pypi release.

**Comment by johngian** [comment]
That said I lean towards

* Restrict PyPI travis deployment on new github releases
* Every time we want a new feature to land to PyPI
  * Change ``__version__``
  * Add ``HISTORY.rst``
  * Merge
  * Draft new github release with the ``__version__`` defined above

**Comment by willkg** [comment]
I think this is a security-related library and as such releases require three properties:

1. releases can only be done by a specific group of people
2. releases are explicit and intentional
3. release tags are signed

I think that might be good enough. However we do it, we're going to want the release process and this library to be vetted by @jvehent and his crew. That's next on my list to coordinate.

Towards that, I think @johngian's process is fine as long as "draft new github release" is really "create a signed tag of the commit to release in the git repository and push it to github". Further, the release process should be documented somewhere.

Beyond that, I don't think it matters a whole lot how it happens as long as it has the above three properties.

Personally, I don't really care about release cadence. Security issues should get released asap. Everything else can be quarterly or whenever. We can release when people need it--that's fine, too. If releasing is too arduous, I'm happy to do it.

**Comment by peterbe** [comment]
Regarding...

1. Does that mean only we 4 humans can do it? And not TravisCI?
2. What does that imply?
3. [A commit can be signed too](https://help.github.com/articles/signing-commits-using-gpg/). I guess that doesn't happen when you press the green merge button in GitHub.

**Comment by jvehent** [comment]
As @willkg said, I think the action of releasing a new version is really issuing a signed git tag, which only a handful a maintainers can do (the 4 humans). When a signed git tag is found, travisci can propagate the release to wherever it needs to (pypi, etc.). [Sops does something similar](https://github.com/mozilla/sops/blob/master/.travis.yml#L23-L33).

Signed commits don't exactly carry the same information as signed tags, which verify a chain of commits, and apply a stamp to it.

**Comment by peterbe** [comment]
> "Signed commits don't exactly carry the same information as signed tags"

That was news to me. I don't think I still understand it but, this answers point 3. Tags it is. 

What you're also saying is that it's ok that Travis goes ahead and makes the deployment (based on the new tag). So that answers point 1 too.

**Comment by jvehent** [comment]
It really goes down to the difference between a commit and a tag. A tag says "I looked at all these commits and decided to release them as a new version". You can't do that with just commits. Signing the tag just adds extra security.

Also, it's impractical to require all commits being signed, because contributors, etc. It's easier to just require tags issued by a handful of maintainers to be signed.

**Comment by johngian** [comment]
I don't think that travis can check if a git tag is signed but we can make sure all version tags are signed. Maybe using ``condition`` [1] we can verify the tag but it needs some more investigation.

Does the following sum up all the above?

* On new feature PRs we change ``HISTORY.rst``
* If devs decide we need a new release
  * Create a commit with a version bump
  * Create a signed tag of the version to that commit
  * Push tag to github
* Travis automatically deploys tags to PyPI
  * We need to make sure that we push only signed tags for version bumps
  * Maybe restrict travis to a specific regex for tags

If we all agree i can document this and set up the automation.

[1] https://docs.travis-ci.com/user/deployment/#Conditional-Releases-with-on%3A

**Comment by peterbe** [comment]
The only nit I can see is "On new feature PRs we change HISTORY.rst". Should be "On new feature or bug fix PRs we change HISTORY.rst".

**Comment by willkg** [comment]
If we forget to change HISTORY in a pr, I'm happy to do it in a separate PR. Generally, it's the "backwards incompatible things" that really needs a note because those affect people upgrading.

**Comment by johngian** [comment]
Fixed after #145
